### PR TITLE
update rclcpp::Waitable API to use references and const

### PR DESCRIPTION
**Pull request by wjwwood**
Splitting this off from https://github.com/ros2/rclcpp/pull/2420 and trying to make the deprecation cycle work.

### CODE REVIEW

**Comment by wjwwood** [comment]
Commit 7d9b19ba43bd36de82819cec1d10d5e07c3bfafa is the main change to the API, and the other commits are fixing the fallout.

**Review by jmachowinski — APPROVED** [review]

**Review by mjcarroll — APPROVED** [review]

**Comment by mjcarroll** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20654/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14983/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21356/)

**Comment by mjcarroll** [comment]
New test failure on this one: 

https://ci.ros2.org/job/ci_linux/20654/testReport/junit/ros2action.ros2action.test/test_cli/test_cli/

**Comment by wjwwood** [comment]
Need to rebase this after https://github.com/ros2/rclcpp/pull/2142 was merged.

**Comment by wjwwood** [comment]
New CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20660/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14991/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21363/)

File: rclcpp/src/rclcpp/guard_condition.cpp [hunk 1]
```diff
@@ -92,19 +92,19 @@ GuardCondition::exchange_in_use_by_wait_set_state(bool in_use_state)
...
 
   if (exchange_in_use_by_wait_set_state(true)) {
-    if (wait_set != wait_set_) {
+    if (&wait_set != wait_set_) {
       throw std::runtime_error("guard condition has already been added to a wait set.");
     }
   } else {
-    wait_set_ = wait_set;
+    wait_set_ = &wait_set;
   }
```
Comment (jmachowinski) [inline rclcpp/src/rclcpp/guard_condition.cpp:105]: General remark, this code is kind of odd, it looks more like debug or assert code. I don't see it as blocker for this RP though.

**Review by jmachowinski — APPROVED** [review]
LGTM

**Pull request merged**